Lovelace: Allow press-and-hold on picture-elements elements.

thomasloven opened the pull request:
This is a native implementation of my [long-press](https://github.com/thomasloven/lovelace-long-press) element. It adds a new `hold_action` option to all picture-elements elements which have a `tap_action` option. The `hold_action` (which takes the same values as `tap_action`) is performed if the element is pressed and held, while the `tap_action` is performed if the element is clicked or tapped.

There are still some problems, but I'm posting this PR hoping for a bit of feedback. Please test it out and let me know what you think of the feel of the hold.

## Example configuration
```yaml
- type: picture-elements
  image: /local/floorplan.png
  elements:
    - type: state-icon
      entity: light.bed_light
      tap_action: toggle
      hold_action: more-info
      style:
        top: 50%
        left: 50%
```

## Current problems

- The hold time is 300 ms. This is rather short, but any longer, and e.g. an iPhone will ask if you wish to save the picture you're holding on.
- The `hold_action` is performed before the mouse button is released. This means that if `hold_action` is `more-info`, it's possible that the more-info-dialog pops up with e.g. the color wheel below the mouse cursor, and you accidentally change a light color.
- `tap_action` and `hold_action` share variables for `navigate` and `call-service`. That means you can't make two different service calls or navigate to two different pages with tap and hold respectively.

## Alternative method
In this implementation, the `hold_action` is performed once the element is held for a short time. Another method could be to perform the action only when the element is held for a while and then released.
This would remove some complications (ugly hack to avoid closing more-info-dialog prematurely, see comments on code), but bring others (prompt to save image on iOS, see above). It might feel less intuitive, and would probably need some kind of indication of when an entity has been pressed long enough.



## Documentation
home-assistant/home-assistant.io#6982

thomasloven committed:
Allow press-and-hold on picture-elements elements.

thomasloven commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -10,13 +10,71 @@ import computeStateName from '../../../common/entity/compute_state_name';
+            .querySelector('home-assistant')
+            .$$('ha-more-info-dialog');
+          moreInfoDialog.setAttribute('modal', true);
+        }
```
Once the mouse button is released after a `hold`, a `mouseup` or `click` event is fired. If the more-info-dialog was opened as `hold_action`, that will immediately close it again. Therefore the dialog is given the attribute `modal`, which means it's not closed by click outside the dialog. The `modal` attribute is removed on `mouseup` above (line 43).

thomasloven commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -10,13 +10,71 @@ import computeStateName from '../../../common/entity/compute_state_name';
+        .querySelector('home-assistant')
+        .$$('ha-more-info-dialog');
+      if (moreInfoDialog) {
+        setTimeout(() => moreInfoDialog.removeAttribute('modal'), 0);
```
This removes the `modal` attribute from the more-info-dialog described in a comment below.
The use of a timeout ensures that the attribute is not removed until after all `click/mouseup` event handlers have been run.

balloob commented:
I think that there are too many current problems. It will be a maintenance nightmare and more things will pop up that will need hacks like the more info attribute.

thomasloven commented:
I absolutely agree (I told you it was very hacky).

I've been experimenting a bit today, and think that the alternative method described above would be a better choice. If the event happens after the mouse button or touch is released, the problem goes away. It also seems the save-image-prompt on IOS can be removed by a css setting. Leaving no major problems with that method.

What remains is the need to indicate that the entity has been held for long enough.
I was thinking something like the material design ink ripple would make sense, but lack the skill to implement it. The effect needs to be big enough to be seen below a large thumb, and should remain on screen until the click/touch is released.

balloob commented:
Yeah alternative seems better.

May I suggest you use mwc-ripple

https://www.npmjs.com/package/@material/mwc-ripple

The documentation is lacking but you can add it by importing it and then rendering `<mwc-ripple>`. When long press is active, just set the primary flag: `<mwc-ripple primary>`. You can force it to be active using `<mwc-ripple active>`

thomasloven commented:
With the indicator, the alternative method feels really good.
But I still haven't gotten the indication quite right using mwc-ripple. It's limited to the size of the element, which is a bit small...

balloob commented:
What if we put an element around the element we're wrapping. The wrapper gets the positioning style applied and we add a 16px padding ?

thomasloven commented:
Then the entire wrapper element will need to be clickable. I tried a 100x100 pixel div around each element (comparable to about 30 px padding around an mdi icon), and that looked pretty good on a touch screen, but didn't feel like it made sense with a mouse since you could trigger the actions far outside the icon. Also, there's a risk that the clickable areas overlap other icons.

But perhaps there should be different approach for touch vs mouse. I'm already checking if touch is available to determine which events to listen to... 
A related thought is that one could use right click with a mouse, and long press only for touch devices.

thomasloven commented:
I added an option to set the time to trigger a long-press with `hold_time: miliseconds` (default is 500), and rebased onto master with the new formating.

With that I'm removing the WIP label. I feel this is actually ready for review.

I'm still not quite satisfied with the ripple thing. It's a bit small and hard to see against some backgrounds, but there may be ways to improve visibility in upcoming versions of mwv-ripple. Also, what's important is the long-pressability anyway...

Should yarn.lock really be included in the commit?

balloob commented:
![image](https://user-images.githubusercontent.com/1444314/47002782-2983bb80-d12e-11e8-9a9c-55566bbe6fdd.png)

You actually _removed_ yarn.lock. That's wrong.

balloob commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -1,8 +1,12 @@
 import computeStateName from "../../../common/entity/compute_state_name";
+import "@material/mwc-ripple";
+
+templateFactory.templateFactory = templateFactory;
```
this is no longer needed if you update to latest version.

thomasloven replied:
Fixed

balloob commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -12,8 +16,68 @@ import computeStateName from "../../../common/entity/compute_state_name";
-      handleClick(hass, config) {
-        const tapAction = config.tap_action || "more-info";
+      registerMouse(config) {
+        const indicator = this.shadowRoot.querySelector("div mwc-ripple");
```
I've realized, we should not patch every element to have a ripple but we should create a ripple when we need it and inject it into the shadow dom.

```js
let ripple = null;

const startAnimation = () => {
  if (ripple === null) {
    ripple = document.createElement('mwc-ripple');
    this.shadowRoot.append(ripple);
  }
  // …
}
```

balloob replied:
That way we're also not going to render 100 ripples on startup and realize we only need 1 when a user clicks on it.

balloob replied:
relevant talk: https://youtu.be/zfQoleQEa4w?t=1373

balloob replied:
(it actually talks about ripples! 😆)

thomasloven replied:
Unfortunately, ripple doesn't seem to work if placed directly in the shadowroot.

thomasloven replied:
I can probably get it to work if all elements have a wrapping div, though.

thomasloven replied:
Fixed. Also looks better when the ripple doesn't happen on short presses too, imo.

thomasloven commented:
I didn't remove yarn.lock, but it seems one line for each pacakge has been removed...

`integrity sha512-....`

Did I do something stupid when I installed new packages?

balloob commented:
Ooh, update your yarn version :)

thomasloven commented:
Thanks!

balloob commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -12,8 +13,78 @@ import computeStateName from "../../../common/entity/compute_state_name";
+        const createRipple = () => {
+          if (ripple === null) {
+            ripple = document.createElement("mwc-ripple");
+            this.shadowRoot.querySelector("div").appendChild(ripple);
```
It feels weird that we have to insert divs into the elements. I think that if we set the elements to render `display: inline-block`, it should be able to ripple in the host.

thomasloven replied:
Here's the problem: https://github.com/material-components/material-components-web-components/blob/master/packages/ripple/src/ripple-directive.ts#L79
It tries to read and set `.style.position` of its parent node, and shadowRoot doesn't have a style property.

And if the mwc-ripple is outside of shadowRoot all kinds of weird things happen...

balloob replied:
You should never hit that code as both are in the same shadow root and thus have same root node?

balloob replied:
Ah, I see that they are passing the holder of the shadow root as interaction node: https://github.com/material-components/material-components-web-components/blob/master/packages/ripple/src/mwc-ripple.ts#L41

So it's not the same. 

What kind of weird things happen ? Did you try setting the host to `display: inline-block` ?

balloob replied:
Actually, we could override interactionNode on the ripple to be the first child in the shadowRoot. That means the root nodes will be the same

balloob replied:
Complicated 😄

thomasloven replied:
If I do
```javascript
this.shadowroot.appendChild(ripple);
console.log(ripple.parentNode);
```
The console shows
```
#shadowRoot (open)
```
followed by

```
Uncaught (in promise) TypeError: Cannot read property 'position' of undefined
at rippleNode (ripple-directive.ts:79)
```

If I do
```
this.appendChild(ripple);
```

The entity jumps around and lands somewhere entirely different as soon as I click it. Adding `display: inline-block` to `:host` doesn't fix it.

thomasloven replied:
> Complicated

Tell me about it!

thomasloven replied:
Overriding interactionNode with `this.shadowRoot.childNodes[2]` seems to work... but that certainly doesn't seem robust.

Child node 0 is `<style>` and child node 1 is `#text`, which I don't even know what it is... And that's just for the *current* elements.

balloob commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -12,8 +13,78 @@ import computeStateName from "../../../common/entity/compute_state_name";
+        };
+
+        const startAnimation = () => {
+          if (ripple) {
```
Don't we have a guaranteed ripple when we come here?

balloob replied:
Also wouldn't it make more sense to just embed createRipple inside this function? No need to do it earlier than that we  want to show it. _Do less and be lazy._

thomasloven replied:
True.

It seems the ripple needs some time to render or something before it can be activated. Perhaps it resolves an `updateComplete` or something that can be used...

balloob commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -12,8 +13,78 @@ import computeStateName from "../../../common/entity/compute_state_name";
+          if (ripple) {
+            ripple.removeAttribute("primary");
+            ripple.removeAttribute("active");
+            ripple.parentNode.removeChild(ripple);
```
I think that instead of removing it, we can just set the attribute `disabled`. Then we don't have to remove all the other attributes either.

thomasloven replied:
I thought so too, but if it's disabled, it just stops animating, i.e. doesn't disappear.

balloob commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -12,8 +13,72 @@ import computeStateName from "../../../common/entity/compute_state_name";
+            ripple.setAttribute("unbounded", "");
+            ripple.setAttribute("primary", "");
+            ripple.setAttribute("active", "");
+            ripple.removeAttribute("disabled");
```
remove this.

balloob commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -12,8 +13,72 @@ import computeStateName from "../../../common/entity/compute_state_name";
+          if (ripple) return;
+          ripple = document.createElement("mwc-ripple");
+          this.shadowRoot.querySelector("div").appendChild(ripple);
+          ripple.updateComplete.then(() => {
```
Why wait to updateComplete? that means it will render twice? Also, instead of attributes you can set properties: `ripple.unbounded = true`

balloob replied:
also, you could use 
```js
async () => {
  await ripple.updateComplete;
  ripple.setAttribute(…)
}

thomasloven replied:
Setting ripple active doesn't work until it has rendered once.

balloob commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -12,8 +13,72 @@ import computeStateName from "../../../common/entity/compute_state_name";
+        const startAnimation = async () => {
+          if (ripple) return;
+          ripple = document.createElement("mwc-ripple");
+          this.shadowRoot.querySelector("div").appendChild(ripple);
```
One final thought. Could we wrap the ripple in a div?

Just really trying to have to avoid an extra div having to be added to each element

balloob reviewed (approved):
this is ok to merge if the final thought doesn't work out or if you don't consider it worth exploring.

thomasloven commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -12,8 +13,84 @@ import computeStateName from "../../../common/entity/compute_state_name";
+        rippleWrapper.style.top = "50%";
+        rippleWrapper.style.left = "50%";
+        rippleWrapper.style.transform = "translate(-50%, -50%)";
+        rippleWrapper.style.pointerEvents = "none";
```
If anyone knows a more handsom way of setting lots of style properties at once I'm all ears.

zsarnett replied:
You could create a class and add/remove the class. Or you could use `Object.assign(rippleWrapper.style, {top: "50%", left: "50%", transform: "translate(-50%, -50%)", pointerEvents: "none"})`

zsarnett replied:
I would recommend the latter

thomasloven replied:
That's what I was looking for.

thomasloven commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -12,8 +13,84 @@ import computeStateName from "../../../common/entity/compute_state_name";
+
+        this.addEventListener(mouseDown, () => {
+          held = false;
+          loadRipple();
```
The ripple needs to be lazy-loaded on mouse-down rather than when the animation happens.
The ripple can't be set active until it has finished rendering, and for some reason waiting for `ripple.updateComplete` doesn't work on iphone...

thomasloven commented on src/panels/lovelace/mixins/element-click-mixin.js:
```diff
@@ -12,8 +13,84 @@ import computeStateName from "../../../common/entity/compute_state_name";
+        this.shadowRoot.appendChild(rippleWrapper);
+        rippleWrapper.style.position = "absolute";
+        rippleWrapper.style.width = "100px";
+        rippleWrapper.style.height = "100px";
```
This might be a bit too much on a computer with a mouse...
Would it make sense to have bigger ripples on a touch screen, or do we want to keep the interface absolutely consistent?

thomasloven replied:
After trying it out; screw consistency. It's grotesquely big on a computer. I'll fix this.

thomasloven committed:
Different size ripples for touch and no-touch

thomasloven committed:
Formating change
